DEPRECATION.md

```diff
@@ -0,0 +1,100 @@
+    * Useful new feature
+
+  * __High impact__
+
+    * Major bugs
+
+3. If the impact merits the cost, let's try to make the change in a backward compatible manner.
+
+  * Options include:
+    * When deprecating, say, a class, leave a shim to its replacement.
+      * E.g., an existing class implementation might be replaced by subclassing its more general replacement.
+    * A function signature can likely be changed with a keyword argument.
+    * You can decorate with `Registrable.register` multiple times.
+    * Features can be hidden behind flags.
+      * E.g., simple boolean arguments to constructors.
+    * Files can simply be forked. (for extreme cases)
+      * This could occur when extending an existing model, for instance.
+      * Copying a file, modifying it, and then de-duplicating can be simpler than parameterizing existing code to handle a new use-case in a purely backwards compatible manner.
+
+  * Likely no deprecation needs to occur in this case. However, in the spirit of providing a single preferred solution, one may still mark the old component as deprecated while _not_ removing it. Usages in AllenNLP should be removed so that our warnings are not spammy. Though silencing warnings is an option, we should "eat our own dog food" or we can hardly expect our users to migrate.
+
+4. If no backward compatible change can be made, it's adivsable to first consult with other developers.
+
+  * Ideally they will propose an alternative solution or help mitigate the churn during the change.
+  * Next proceed to __Mechanics__.
+
+## Mechanics
+
+1. Mark the old component as deprecated by adding a `DeprecationWarning`.
+  * This should include whether the component will be removed.
+  * If it must be removed, describe why in an accompanying comment. Link relevant issues.
+  * [Example](https://github.com/allenai/allennlp/blob/cb9651a4c77c10cbd2d76f79b85c6453386dc229/allennlp/modules/text_field_embedders/basic_text_field_embedder.py#L141)
+  * Provide the version and date when we will remove the feature if applicable.
+    * We should support whichever is longest.
+    * The code should live for at least one full minor version and 3 months before removal.
+      * e.g., if you're committing the deprecation to master while version 0.8.4 is out, then it should live throughout version 0.9 and can first be removed in version 0.10.0.
+      * In particular, this should be at least a minor release, i.e. m.n.0.
+      * If this isn't possible, consult with other developers. You should have a compelling rationale.
+
+2. Remove any AllenNLP usages of the deprecated feature to avoid warnings.
+  * Suppressing warnings should be done rarely. [See here for instance.](https://github.com/allenai/allennlp/blob/9719b5c71207e642276fb1209ea1a4c8467e0792/allennlp/modules/token_embedders/embedding.py#L14)
+
+3. Create a Github issue for the actual removal and wait for the requisite removal release.
```

Comment:
I think we need to associate the GitHub issue with a milestone corresponding to the future release in which you would like the component to be removed.  Otherwise we'll lose track of it.

Reply:
I'm somewhat conflicted here. We don't currently appear to have any future milestones. So this would mean the deprecator would need to create a new release milestone. Are we really using them extensively? I don't want to create potential busywork.

My counterpoint would be that if no one is driving the removal forward, it's clearly not urgent. Our periodic issue sweeps would then seem sufficient. (In other words, if we intend to deprecate in version 0.10.0, but actually deprecate in version 0.11.0, that's a totally acceptable outcome.)

Reply:
@brendan-ai2 we've used milestones up to the 0.8.0 release.  They're often invaluable for making sure we tie up loose ends before a release.  Often there are a number of issues / PRs people want merged before a release is cut--and this is an effective method of managing them.

https://github.com/allenai/allennlp/milestones?state=closed

Reply:
I think my point is we should continue to use milestones to manage releases.  I think your point is that we don't necessarily need to put deprecations on milestones, which is fine.  We should make the "will be removed on or after X" instead of the "will be removed in X" as we have been in deprecation warnings.